src/mpl/include/mpl_atomic.h

```diff
@@ -0,0 +1,95 @@
+static inline uint32_t MPL_atomic_acquire_load_uint32 (const MPL_atomic_uint32_t * ptr);
+static inline void MPL_atomic_relaxed_store_uint32 (MPL_atomic_uint32_t * ptr, uint32_t val);
+static inline void MPL_atomic_release_store_uint32 (MPL_atomic_uint32_t * ptr, uint32_t val);
+static inline uint32_t MPL_atomic_cas_uint32 (MPL_atomic_uint32_t * ptr, uint32_t oldv, uint32_t newv);
+static inline uint32_t MPL_atomic_swap_uint32 (MPL_atomic_uint32_t * ptr, uint32_t val);
+static inline uint32_t MPL_atomic_fetch_add_uint32 (MPL_atomic_uint32_t * ptr, uint32_t val);
+static inline uint32_t MPL_atomic_fetch_sub_uint32 (MPL_atomic_uint32_t * ptr, uint32_t val);
+static inline int64_t MPL_atomic_relaxed_load_int64 (const MPL_atomic_int64_t * ptr);
+static inline int64_t MPL_atomic_acquire_load_int64 (const MPL_atomic_int64_t * ptr);
+static inline void MPL_atomic_relaxed_store_int64 (MPL_atomic_int64_t * ptr, int64_t val);
+static inline void MPL_atomic_release_store_int64 (MPL_atomic_int64_t * ptr, int64_t val);
+static inline int64_t MPL_atomic_cas_int64 (MPL_atomic_int64_t * ptr, int64_t oldv, int64_t newv);
+static inline int64_t MPL_atomic_swap_int64 (MPL_atomic_int64_t * ptr, int64_t val);
+static inline int64_t MPL_atomic_fetch_add_int64 (MPL_atomic_int64_t * ptr, int64_t val);
+static inline int64_t MPL_atomic_fetch_sub_int64 (MPL_atomic_int64_t * ptr, int64_t val);
+static inline uint64_t MPL_atomic_relaxed_load_uint64 (const MPL_atomic_uint64_t * ptr);
+static inline uint64_t MPL_atomic_acquire_load_uint64 (const MPL_atomic_uint64_t * ptr);
+static inline void MPL_atomic_relaxed_store_uint64 (MPL_atomic_uint64_t * ptr, uint64_t val);
+static inline void MPL_atomic_release_store_uint64 (MPL_atomic_uint64_t * ptr, uint64_t val);
+static inline uint64_t MPL_atomic_cas_uint64 (MPL_atomic_uint64_t * ptr, uint64_t oldv, uint64_t newv);
+static inline uint64_t MPL_atomic_swap_uint64 (MPL_atomic_uint64_t * ptr, uint64_t val);
+static inline uint64_t MPL_atomic_fetch_add_uint64 (MPL_atomic_uint64_t * ptr, uint64_t val);
+static inline uint64_t MPL_atomic_fetch_sub_uint64 (MPL_atomic_uint64_t * ptr, uint64_t val);
+static inline void *MPL_atomic_relaxed_load_ptr (const MPL_atomic_ptr_t * ptr);
+static inline void *MPL_atomic_acquire_load_ptr (const MPL_atomic_ptr_t * ptr);
+static inline void MPL_atomic_relaxed_store_ptr (MPL_atomic_ptr_t * ptr, void *val);
+static inline void MPL_atomic_release_store_ptr (MPL_atomic_ptr_t * ptr, void *val);
+static inline void *MPL_atomic_cas_ptr (MPL_atomic_ptr_t * ptr, void *oldv, void *newv);
+static inline void *MPL_atomic_swap_ptr (MPL_atomic_ptr_t * ptr, void *val);
+
+static inline void MPL_atomic_write_barrier(void);
+static inline void MPL_atomic_read_barrier(void);
+static inline void MPL_atomic_read_write_barrier(void);
+static inline void MPL_atomic_compiler_barrier(void);
+
+#if defined(MPL_USE_UNSAFE_PRIMITIVES)
+#include "mpl_atomic_unsafe.h"
```

Comment:
(I never liked the word `unsafe`, anyway--) how about `#include `atomics/mpl_atomic_unsafe.h`, by moving the secondary header files into a subfolder?

Reply:
As far as I checked, mpl doesn't like to create subdirectories; for example, all types `mpl_thread_xxx.h` headers are directly put into `src/mpl/include/` (see  https://github.com/pmodels/mpich/tree/master/src/mpl/include). For better readability, I can move `mpl_atomic_xxx.h` files to a subdirectory (if this is allowed), or I can make another PR later that moves `mpl_atomic_xxx.h`, `mpl_thread_xxx.h`, `mpl_shm_xxx.h`, and  `mpl_timer_xxx.h` to associated subdirectories (e.g., `src/mpl/include/atomic/`, `src/mpl/include/thread/`, ...).